crates/accelerate/src/commutation_checker.rs

```diff
@@ -0,0 +1,725 @@
+    (SmallVec<[ParameterKey; 3]>, SmallVec<[ParameterKey; 3]>),
+);
+// Need a struct instead of a type definition because we cannot implement serialization traits otherwise
+#[derive(Clone, Debug)]
+struct CommutationCacheEntry {
+    mapping: HashMap<CacheKey, bool>,
+}
+impl CommutationCacheEntry {
+    fn get(&self, key: &CacheKey) -> Option<&bool> {
+        self.mapping.get(key)
+    }
+    fn iter(&self) -> Iter<'_, CacheKey, bool> {
+        self.mapping.iter()
+    }
+
+    fn insert(&mut self, k: CacheKey, v: bool) -> Option<bool> {
+        self.mapping.insert(k, v)
+    }
+}
+
+impl ToPyObject for CommutationCacheEntry {
+    fn to_object(&self, py: Python) -> PyObject {
+        let out_dict = PyDict::new_bound(py);
+        for (k, v) in self.iter() {
+            let qubits = PyTuple::new_bound(py, k.0.iter().map(|q| q.map(|t| t.0)));
+            let params0 = PyTuple::new_bound(py, k.1 .0.iter().map(|pk| pk.0));
+            let params1 = PyTuple::new_bound(py, k.1 .1.iter().map(|pk| pk.0));
+            out_dict
+                .set_item(
+                    PyTuple::new_bound(py, [qubits, PyTuple::new_bound(py, [params0, params1])]),
+                    PyBool::new_bound(py, *v),
+                )
+                .expect("Failed to construct commutation cache for serialization");
+        }
+        out_dict.into_any().unbind()
+    }
+}
+
+type CacheKeyRaw = (
+    SmallVec<[Option<u32>; 2]>,
+    (SmallVec<[f64; 3]>, SmallVec<[f64; 3]>),
+);
+impl<'py> FromPyObject<'py> for CommutationCacheEntry {
+    fn extract_bound(b: &Bound<'py, PyAny>) -> Result<Self, PyErr> {
+        let dict = b.downcast::<PyDict>()?;
+        let mut ret = hashbrown::HashMap::with_capacity(dict.len());
+        for (k, v) in dict {
+            let raw_key: CacheKeyRaw = k.extract()?;
+            let qubits = raw_key.0.iter().map(|q| q.map(Qubit)).collect();
+            let params0: SmallVec<_> = raw_key.1 .0.iter().map(|p| ParameterKey(*p)).collect();
+            let params1: SmallVec<_> = raw_key.1 .1.iter().map(|p| ParameterKey(*p)).collect();
+            let v: bool = v.extract()?;
+            ret.insert((qubits, (params0, params1)), v);
+        }
+        Ok(CommutationCacheEntry { mapping: ret })
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+struct ParameterKey(f64);
+
+impl ParameterKey {
+    fn key(&self) -> u64 {
+        self.0.to_bits()
+    }
+}
+
+impl std::hash::Hash for ParameterKey {
+    fn hash<H>(&self, state: &mut H)
+    where
+        H: std::hash::Hasher,
+    {
+        self.key().hash(state)
+    }
+}
+
+impl PartialEq for ParameterKey {
+    fn eq(&self, other: &ParameterKey) -> bool {
+        self.key() == other.key()
+    }
+}
+
+impl Eq for ParameterKey {}
+
+fn hashable_params(params: &[Param]) -> SmallVec<[ParameterKey; 3]> {
+    params
+        .iter()
+        .map(|x| {
+            if let Param::Float(x) = x {
+                ParameterKey(*x)
+            } else {
+                panic!()
```

Comment:
While not explicitly necessary, I'd suggest leaving a panic message here. Something like:

```suggestion
                panic!("Can't hash a non-float instruction parameter")
```